validator_worker/src/sentry_interface.rs

```diff
@@ -1,143 +1,261 @@
+            let auth_token = self
+                .adapter
+                .get_auth(&validator)
+                .expect("Failed to get user auth token");
+            match propagate_to(
+                &auth_token,
+                self.config.propagation_timeout,
+                &validator,
+                &serialised_messages,
+            ) {
+                Ok(_) => return,
+                Err(e) => handle_http_error(e, &validator.url),
+            }
+        }
     }
 
-    async pub fn get_latest_msg(from: &str, type: &str) -> impl Future<Output = Result<Vec<Channel>, Error>>  {
+    pub async fn get_latest_msg(
+        &self,
+        from: String,
+        message_type: String,
+    ) -> Result<ValidatorMessageResponse, reqwest::Error> {
         let future = self
             .client
-            .get(format!("{}/validator-messages/{}/{}?limit=1", self.sentry_url, from, type))
+            .get(&format!(
+                "{}/validator-messages/{}/{}?limit=1",
+                self.validator_url, from, message_type
+            ))
             .send()
-            .and_then(|mut res: Response| res.json::<ChannelAllResponse>());
+            .and_then(|mut res: Response| res.json::<ValidatorMessageResponse>())
+            .compat();
 
-        await!(future.compat())
+        await!(future)
     }
 
-    async pub fn get_our_latest_msg(type: &str) -> impl Future<Output = Result<Vec<Channel>, Error>>  {
-        let whoami = adapter.whoami();
-        self.get_latest_msg(whoami, type)
+    pub async fn get_our_latest_msg(
+        &self,
+        message_type: String,
+    ) -> Result<ValidatorMessageResponse, reqwest::Error> {
+        let whoami = self.adapter.whoami();
+        await!(self.get_latest_msg(whoami, message_type))
+    }
+
+    pub async fn get_last_approved(&self) -> Result<LastApprovedResponse, reqwest::Error> {
+        let future = self
+            .client
+            .get(&format!("{}/last-approved", self.validator_url))
+            .send()
+            .and_then(|mut res: Response| res.json::<LastApprovedResponse>());
+
+        await!(future.compat())
     }
 
-    pub fn get_last_approved() -> impl Future<Output = Result<Vec<Channel>, Error>> {
+    pub async fn get_last_msgs(&self) -> Result<LastApprovedResponse, reqwest::Error> {
         let future = self
             .client
-            .get(format!("{}/last-approved", self.sentry_url))
+            .get(&format!(
+                "{}/last-approved?withHearbeat=true",
+                self.validator_url
+            ))
             .send()
-            .and_then(|mut res: Response| res.json::<ChannelAllResponse>());
+            .and_then(|mut res: Response| res.json::<LastApprovedResponse>());
 
         await!(future.compat())
     }
 
-    pub fn get_last_msgs() {
-         let future = self
+    pub async fn get_event_aggregates(
+        &self,
+        after: DateTime<Utc>,
```

Comment:
Now I am not sure how correct is the JS implementation and how it makes sense here, but in the Producer if there is no previous `Accounting` message, it will create one with `lastEvAggr: new Date(0)` (i.e. `Jan 01 1970`) and it will use it for the `after`. Maybe we do need the after to be an Option after all?

@Ivshti @samparsky

Reply:
this is a question.